cmake/MacroGuardTest.cmake

```
@@ -0,0 +1,107 @@
+# PROJECT_NAME: bsoncxx or mongocxx.
+# GUARDED_MACROS: list of macros that should be guarded by prelude/postlude headers.
+# INCLUDE_FILTERS: list of regex filters for headers to be added to the tests (must be relative to PROJECT_SOURCE_DIR).
+# EXCLUDE_FILTERS: list of regex filters for headers to be excluded (applied after INCLUDE_FILTERS).
+# PROJECT_TEST_PROPERTIES: the (bsoncxx|mongocxx)_test_properties target to be linked with.
+function(add_macro_guard_test)
+    set(opt_args "")
+    set(single_args "PROJECT_NAME;PROJECT_TEST_PROPERTIES")
+    set(multi_args "GUARDED_MACROS;INCLUDE_FILTERS;EXCLUDE_FILTERS")
+
+    cmake_parse_arguments(PARSED "${opt_args}" "${single_args}" "${multi_args}" ${ARGN})
+
+    if(NOT "${PARSED_UNPARSED_ARGUMENTS}" STREQUAL "")
+        message(FATAL_ERROR "unrecognized argument: ${PARSED_UNPARSED_ARGUMENTS}")
+    endif()
+
+    foreach(required_arg PROJECT_NAME PROJECT_TEST_PROPERTIES GUARDED_MACROS INCLUDE_FILTERS)
+        if("${PARSED_${required_arg}}" STREQUAL "")
+            message(FATAL_ERROR "missing required argument: ${required_arg}")
+        endif()
+    endforeach()
+
+    set(GUARDED_HEADERS "")
+
+    foreach(filter ${PARSED_INCLUDE_FILTERS})
+        file(GLOB_RECURSE headers
+            LIST_DIRECTORIES false
+            RELATIVE ${PROJECT_SOURCE_DIR}
+            "${PROJECT_SOURCE_DIR}/${filter}"
+        )
+        list(APPEND GUARDED_HEADERS ${headers})
+    endforeach()
```

You can pass multiple glob patterns to `file(GLOB)` in a single call:

```suggestion
    # Make the include patterns absolute from the project's source dir:
    list(TRANSFORM PARSED_INCLUDE_FILTERS PREPEND "${PROJECT_SOURCE_DIR}/")
    # Collect included headers:
    file(GLOB_RECURSE headers CONFIGURE_DEPENDS RELATIVE "${PROJECT_SOURCE_DIR}" ${PARSED_INCLUDE_FILTERS})
```